### PR TITLE
Add the event: WorldSoundEvent

### PR DESCRIPTION
**Pull request by aderoian**
## Introduction
Adds an API event to listen for when a Sound is played to a world using World->addSound(). This event can be cancelled, or the sound, position, and/or target players can be changed.

### Relevant issues
#5318 

### API changes
Addition of the event: ``WorldSoundEvent``

### Behavioural changes
n/a

## Backwards compatibility
n/a

## Follow-up
n/a

## Tests
Test plugin to test cancelling, changing the sound, changing the location, and changing the target players.
```php
final class Main extends PluginBase implements Listener{

	public function onEnable() : void{
		$this->getServer()->getPluginManager()->registerEvents($this, $this);
	}

	/**
	 * @priority MONITOR
	 * @ignoreCancelled true
	 */
	public function onWorldSound(WorldSoundEvent $event) : void{
		// $event->cancel(); // No sounds are played (works)

		// $event->setSound(new AnvilFallSound()); // All sounds are replaced by anvil sounds (works)

		// $event->setPosition($event->getPosition()->add(0, 25, 0)); // Sound is quieter (works)

		// $event->setPlayers([]); // Sound is sent to no one (works)
	}
}
```

### CODE REVIEW

File: src/event/world/WorldSoundEvent.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+use pocketmine\event\Cancellable;
+use pocketmine\event\CancellableTrait;
+use pocketmine\math\Vector3;
+use pocketmine\player\Player;
+use pocketmine\world\sound\Sound;
+use pocketmine\world\World;
+
+/**
+ * Called when a sound is played in a world using World->playSound()
```
Comment (dktapps) [inline src/event/world/WorldSoundEvent.php]:
```suggestion
 * Called when a sound is played in a world
```
This is already covered by the @see below

File: src/event/world/WorldSoundEvent.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	use CancellableTrait;
+
+	/**
+	 * @param Player[]|null $players
+	 */
+	public function __construct(
+		World $world,
+		private Vector3 $position,
+		private Sound $sound,
+		private ?array $players = null
```
Comment (dktapps) [inline src/event/world/WorldSoundEvent.php]: why does this accept null?

Comment (aderoian) [reply src/event/world/WorldSoundEvent.php]: https://github.com/pmmp/PocketMine-MP/blob/stable/src/world/World.php#L661-L667

If players == null the world just sends it to the viewers. Developers may only want to change the sound and not the target, so why make them have to figure out who to send that sound to.

Comment (IvanCraft623) [reply src/event/world/WorldSoundEvent.php]: I don't like that this can be null, it can be understood as that no player will hear the sound.

Comment (aderoian) [reply src/event/world/WorldSoundEvent.php]: What do we think about a field in the event such as ``toViewersOnly`` which essentially acts as if $players was null?

Comment (dktapps) [reply src/event/world/WorldSoundEvent.php]: The event would already provide the viewers, so your point doesn't make any sense.

Comment (aderoian) [reply src/event/world/WorldSoundEvent.php]: Currently, if no players are specified (``$players = null``) the server adds them to a packet buffer, and does not actually get the viewers of that chunk at that exact moment.

Comment (dktapps) [reply src/event/world/WorldSoundEvent.php]: Hmm, I see. That means that collecting the viewers at that point would change the internal buffering behaviour of sending sounds. Not sure what the best solution for that is.

Comment (aderoian) [reply src/event/world/WorldSoundEvent.php]: The solution that I used was if null was passed get the players near the sound before sending the event. The buffering behavior gets the players who can see that chunk either way.

**Comment by dktapps** [comment]
I'm doubtful that allowing the position of the sound to be changed is useful or desirable.

**Comment by aderoian** [comment]
Yeah, I can agree with that.

**Comment by aderoian** [comment]
Any chance we can change this PR to also do the same thing for ``World::addParticle()``?

**Comment by dktapps** [comment]
Better keep that for a separate PR.

**Comment by aderoian** [comment]
What else is this pr waiting on?

**Comment by dktapps** [comment]
> I'm doubtful that allowing the position of the sound to be changed is useful or desirable.

**Review by aderoian — COMMENTED** [review]
Is it just me or does ``$players`` never get updated from the event, therefore the check for whether the array is changed is always `true`?

**Comment by JavierLeon9966** [comment]
Yeah, you're right

File: src/world/World.php [hunk 3]
```diff
@@ -677,7 +677,7 @@ public function addSound(Vector3 $pos, Sound $sound, ?array $players = null) : v
 
 		$pk = $ev->getSound()->encode($pos);
 		if(count($pk) > 0){
-			if($players === $this->getViewersForPosition($pos)){
+			if($players === $ev->getRecipients()){
```
Comment (dktapps) [inline src/world/World.php]: This is not the correct fix. The correct solution would be to overwrite $players with the result of getRecipients() before entering this code.

**Review by dktapps — APPROVED** [review]

**Pull request merged**